utils/rand/rand.go

```diff
@@ -0,0 +1,169 @@
+package rand
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"errors"
+	"fmt"
+)
+
+// This package is a wrppaer around true RNG crypto/rand.
+// It implements useful tools using the true RNG and that
+// are not exported by the crypto/rand package.
+// This package does not implement any determinstic RNG (Pseudo RNG)
+// unlike the package flow-go/crypto/random.
+
+var randFailure = errors.New("crypto/rand failed")
+
+// returns a random uint64
+func Uint64() (uint64, error) {
+	buffer := make([]byte, 8) // TODO: declare as a global variable and add a lock?
+	if _, err := rand.Read(buffer); err != nil {
+		return 0, randFailure
+	}
+	r := binary.LittleEndian.Uint64(buffer)
+	return r, nil
+}
+
+// returns a random uint64 strictly less than n
+// errors if n==0
+func Uint64n(n uint64) (uint64, error) {
+	if n == 0 {
+		return 0, fmt.Errorf("n should be strictly positive, got %d", n)
+	}
+	// the max returned random is n-1 > 0
+	max := n - 1
+	// count the bytes size of max
+	size := 0
+	for tmp := max; tmp != 0; tmp >>= 8 {
+		size++
+	}
+	buffer := make([]byte, 8) // TODO: declare as a global variable and add a lock?
+	// get the bit size of max
+	mask := uint64(0)
+	for max&mask != max {
+		mask = (mask << 1) | 1
+	}
+
+	// Using 64 bits of random and reducing modulo n does not guarantee a high uniformity
+	// of the result.
+	// For a better uniformity, loop till a sample is less or equal to `max`.
+	// This means the function might take longer time to output a random.
+	// Using the size of `max` in bits helps the loop end earlier (the algo stops after one loop
+	// with more than 50%)
+	// a different approach would be to pull at least 128 bits from the random source
+	// and use big number modular reduction by `n`.
+	random := n
+	for random > max {
+		if _, err := rand.Read(buffer[:size]); err != nil {
```

Comment:
you need to either check for numRead (aka `_` in your code), or use io.ReadFull (same elsewhere)

Reply:
In the case of `crypto/rand`, checking one output is enough, as per the doc:
```
// Read is a helper function that calls Reader.Read using io.ReadFull.
// On return, n == len(b) if and only if err == nil.
```
I'll add a comment to clarify.